### ISSUE TITLE
Proposal: Union of Worlds

### ISSUE THREAD

**Issue by Mossaka** [WebAssembly/component-model#169]
I want to raise this issue to propose a new syntax for the union of Worlds. The primary motivation for this proposal is that we want to be able to form a more compresive world by unionizing / combining multiple worlds together to form a bigger one. Concretly, this is motivated by the [`wasi-cloud-core`](https://github.com/WebAssembly/WASI/issues/520) proposal, which tries to define a new World that includes all the wasi-cloud worlds as to provide a feature set that allows a serverless / edge function to do state management or message exchanging. 

## Proposal

This proposal adds a new syntax `include <world-path>` in World definition. Below are a few examples/sketches showing how the `include` syntax would work. 

### Union of two Worlds

```fsharp
// worlds.wit
world my-world-1 {
    import a: self.a
    import b: self.b
    export c: self.c
}

world my-world-2 {
    import foo: self.foo
    import bar: self.bar
    export baz: self.baz
}

world union-my-world {
     include self.my-world-1
     include self.my-world-2
}

is equivalent to

world union-my-world {
    import a: self.a
    import b: self.b
    export c: self.c
    import foo: self.foo
    import bar: self.bar
    export baz: self.baz
}
```

### Include a World that has package and inline imports

> Notice that the inline export has been "copied" over to the World that includes this World. 

```fsharp
// b.wit
interface b { ... }

// a.wit
interface a { ... }

world my-world-1 {
    import a: self.a 
    import b: pkg.b
    import c: io.c // external package
    export d: interface exp { ... }
}

// union.wit

world union-my-world-1 {
    include pkg.my-world-1
}

is equivalent to

world union-my-world-1 {
    import a: pkg.a
    import b: pkg.b
    import c: io.c
    export d: interface exp { ... }
}
```
### Name Conflicts

This is a more challenging example where the two Worlds being included have name conflicts. The solution to this is to use `with` syntax and aliasing to avoid conflicts of names. The `with` syntax is only used to avoid name conflicts for those names that are in conflict with another World. The names that are not in conflicts should stay the same and need not to be specified in the `with` scope.

```fsharp
// my-world-1.wit
world my-world-1 {
    import a: self.a
    import b: self.b
    export d: self.d
}

// my-world-2.wit
world my-world-2 {
    import a: self.a
    import b: self.b
    export c: self.c
}

// union.wit
world union-my-world {
    include pkg.my-world-1 with { a as a1, b as b1 }
    include pkg.my-world-2
}

is equivalent to:

world union-my-world {
    // resolve conflicts
    import a1: pkg.my-world-1.a
    import b1: pkg.my-world-1.b
    export d: pkg.my-world-1.d

    import a: pkg.my-world-2.a
    import b: pkg.my-world-2.b
    export c: pkg.my-world-2.c
}
```

### De-duplication

> When imports and exports of union-ing of multi-worlds have the same definitions, they will be deduplicated. 

```fsharp
// a.wit
// b.wit
// c.wit

// my-world-1.wit
world my-world-1 {
    import a: pkg.a
    import b: pkg.b
    export c: pkg.c
}

// my-world-2.wit
world my-world-2 {
    import a: pkg.a
    import b: pkg.b
    export c: pkg.c
}

// union.wit
world union-my-world {
    include pkg.my-world-1
    include pkg.my-world-2
}

is equivalent to 

world union-my-world {
    import a: pkg.a
    import b: pkg.b
    export c: pkg.c
}
```

Let me know your thoughts on this design! @lukewagner @fibonacci1729 @alexcrichton

**Comment by alexcrichton** [comment]
What you've outlined looks all good to me, although one more case to consider is type definitions in worlds:

```
world foo {
    type a = u32;
}

world bar {
    type a = u32;
}

world baz {
    include self.foo;
    include self.bar; // should this error?
}
```

I think the answer is "yes" it should error which would require the disambiguation syntax you've proposed to resolve. 

I also think that https://github.com/WebAssembly/component-model/pull/164 makes the disambiguation syntax nicer here since there's only one namespace to disambiguate and you don't have to specify either imports or exports.

**Comment by Mossaka** [comment]
Thank you, @alexcrichton and @lukewagner for your comments!

> I also think that https://github.com/WebAssembly/component-model/pull/164 makes the disambiguation syntax nicer here since there's only one namespace to disambiguate and you don't have to specify either imports or exports.

This is really nice!

> To my understanding, this is analogous to the de-duping of two imports of the same interface.

I was thinking the same. Maybe it makes more sense for the following type definitions in worlds to be resolved using `with`

```go
world foo {
    type a = u64;
}

world bar {
    type a = u32;
}

world baz {
    include self.foo with { a as a1}
    include self.bar; 
}
```

> a component targeting the included world won't be a subtype of the unioned world

Ah right, it will break the subtyping rule if we are not careful about union-ing exports. Good point! I agree that for now we can let the host to implicitly treat all exports as optional.

**Comment by alexcrichton** [comment]
That's true yeah, although the subtyping/equality check is currently only present in the wasm binary validator rather than the WIT parser phase, so I'd be tempted to say that to be conservative as a starting point types are required to be disambiguated and perhaps in the future we can relax the rule for structurally equivalent types.

Otherwise though @Mossaka yeah I think the syntax works the same with similar semantics, it's just that types are something to consider in addition to named interfaces/functions.

**Comment by Mossaka** [comment]
> so I'd be tempted to say that to be conservative as a starting point types are required to be disambiguated and perhaps in the future we can relax the rule for structurally equivalent types.

That makes sense. For now, let's disambiguate type defs even though structurally they are equivalent. In this case, does my de-duplication section still makes sense? It might need to check if the duplicated types have the same structures which is equivalent to subtyping/equality check in WIT parser phase.

**Comment by alexcrichton** [comment]
Yeah what you mentioned makes sense to me which is to always require explicit renaming on conflicts.

**Comment by lukewagner** [comment]
I think this is resolved by #174 (but please reopen if there is still more to the proposal to be done).
